File: src/lexer.py
from typing import NoReturn, Union, List

from src.commom import (
    Token,
    RESERVED_KEYWORDS,
    EOF,
    LPAREN,
    MINUS,
    MUL,
    RPAREN,
    PLUS,
    ASSIGN,
    DOT,
    ID,
    SEMI,
    COLON,
    COMMA,
    FLOAT_DIV,
    INTEGER_CONST,
    REAL_CONST
)


class Lexer:
    def __init__(self, input_code: str) -> None:
        self.tokens: List[Token] = []
        self.input_code: str = input_code
        self.pos: int = 0
        self.current_char: Union[str, None] = self.input_code[self.pos]

    def get_tokens(self) -> List[Token]:
        while self.current_char is not None:

            if self.current_char.isspace():
                self._skip_whitespace()
                continue

            if self.current_char == '{':
                self._advance()
                self._skip_comment()
                continue

            if self.current_char.isalpha():
                self.tokens.append(self._id())
                continue

            if self.current_char.isdigit():
                self.tokens.append(self._create_number())
                continue

            if self.current_char == ":" and self._peek() == "=":
                self._advance()
                self._advance()
                self.tokens.append(Token(ASSIGN, ":="))
                continue

            if self.current_char == ";":
                self._advance()
                self.tokens.append(Token(SEMI, ";"))
                continue

            if self.current_char == "+":
                self._advance()
                self.tokens.append(Token(PLUS, "+"))
                continue

            if self.current_char == "-":
                self._advance()
                self.tokens.append(Token(MINUS, "-"))
                continue

            if self.current_char == "*":
                self._advance()
                self.tokens.append(Token(MUL, "*"))
                continue

            if self.current_char == "(":
                self._advance()
                self.tokens.append(Token(LPAREN, "("))
                continue

            if self.current_char == ")":
                self._advance()
                self.tokens.append(Token(RPAREN, ")"))
                continue

            if self.current_char == ".":
                self._advance()
                self.tokens.append(Token(DOT, "."))
                continue

            if self.current_char == ':':
                self._advance()
                self.tokens.append(Token(COLON, ':'))
                continue

            if self.current_char == ',':
                self._advance()
                self.tokens.append(Token(COMMA, ','))
                continue

            if self.current_char == '/':
                self._advance()
                self.tokens.append(Token(FLOAT_DIV, '/'))
                continue

            print('----- ERROR -----')
            print(self.input_code[self.pos - 10:self.pos + 10])
            print('----------')
            self._error()

        self.tokens.append(Token(EOF, None))

        return self.tokens

    def _error(self) -> NoReturn:
        raise Exception("Invalid character")

    def _advance(self) -> None:
        self.pos += 1

        if self.pos > len(self.input_code) - 1:
            self.current_char = None
        else:
            self.current_char = self.input_code[self.pos]

    def _skip_whitespace(self) -> None:
        while self.current_char is not None and self.current_char.isspace():
            self._advance()

    def _create_number(self) -> Token:
        result: str = ""

        while self.current_char is not None and self.current_char.isdigit():
            result += self.current_char
            self._advance()

        if self.current_char == '.':
            result += self.current_char
            self._advance()

            while (
                self.current_char is not None and
                self.current_char.isdigit()
            ):
                result += self.current_char
                self._advance()

            token = Token(REAL_CONST, float(result))  # type: ignore
        else:
            token = Token(INTEGER_CONST, int(result))

        return token

    def _peek(self) -> Union[str, None]:
        peek_pos = self.pos + 1
        if peek_pos > len(self.input_code) - 1:
            return None
        else:
            return self.input_code[peek_pos]

    def _id(self) -> Token:
        result = ""

        while self.current_char is not None and self.current_char.isalnum():
            result += self.current_char
            self._advance()

        return RESERVED_KEYWORDS.get(result, Token(ID, result))

    def _skip_comment(self):
        while self.current_char != '}':
            self._advance()
        self._advance()

File: src/modules/lexer.ts
import { Token } from '../entities';
import { Keywords, TokenTypes, alphanumeric, letters, numbers } from '../types';

/**
 * The lexer is responsible for reading the input code and identifying the tokens in it.
 * It's also responsible for raising an error if the current character is not recognized.
 * @see https://ruslanspivak.com/lsbasi-part3/
 */
export class Lexer {
  private code: string = '';
  private carretPosition: number = 0;
  private currentCharacter: string | undefined;

  constructor(code: string) {
    this.code = code.trim();
    this.currentCharacter = this.code[this.carretPosition];
  }

  /**
   * Returns the next token in the input code.
   * Throws an error if the current character is not recognized.
   */
  getNextToken(): Token {
    while (this.currentCharacter) {
      if (this.currentCharacter === ' ') {
        this.skipWhitespace();
        continue;
      }

      if (this.currentCharacter === '\n') {
        this.skipNewLine();
        continue;
      }

      if (this.currentCharacter === '{') {
        this.advance();
        this.skipComment();
        continue;
      }

      if (letters.includes(this.currentCharacter)) {
        return this.id();
      }

      if (numbers.includes(this.currentCharacter)) {
        return this.createNumber();
      }

      if (this.currentCharacter === ':' && this.peek() === '=') {
        this.advance();
        this.advance();
        return new Token(TokenTypes.ASSIGN, ':=');
      }

      if (this.currentCharacter === ';') {
        this.advance();
        return new Token(TokenTypes.SEMI, ';');
      }

      if (this.currentCharacter === '+') {
        this.advance();
        return new Token(TokenTypes.PLUS, '+');
      }

      if (this.currentCharacter === '-') {
        this.advance();
        return new Token(TokenTypes.MINUS, '-');
      }

      if (this.currentCharacter === '*') {
        this.advance();
        return new Token(TokenTypes.MUL, '*');
      }

      if (this.currentCharacter === '(') {
        this.advance();
        return new Token(TokenTypes.LPAREN, '(');
      }

      if (this.currentCharacter === ')') {
        this.advance();
        return new Token(TokenTypes.RPAREN, ')');
      }

      if (this.currentCharacter === '.') {
        this.advance();
        return new Token(TokenTypes.DOT, '.');
      }

      if (this.currentCharacter === ':') {
        this.advance();
        return new Token(TokenTypes.COLON, ':');
      }

      if (this.currentCharacter === ',') {
        this.advance();
        return new Token(TokenTypes.COMMA, ',');
      }

      if (this.currentCharacter === '/') {
        this.advance();
        return new Token(TokenTypes.FLOAT_DIV, '/');
      }

      this.raiseError(this.getSnippet());
    }

    return new Token(TokenTypes.EOF, TokenTypes.EOF);
  }

  /**
   * Creates the token that contains the numeric value, which can be integer or real.
   */
  private createNumber(): Token {
    let number = '';
    let token = { token: '' as any };

    const getNumberSequence = (
      onExitCb?: (tokenRef: Record<string, Token>) => void,
    ) => {
      while (this.currentCharacter && numbers.includes(this.currentCharacter)) {
        number += this.currentCharacter;
        this.advance();
      }
      onExitCb && onExitCb(token);
    };

    getNumberSequence(() => {
      if (this.currentCharacter !== '.') {
        token.token = new Token(TokenTypes.INTEGER_CONST, number);
        return;
      }

      getNumberSequence((tokenRef) => {
        number += this.currentCharacter;
        this.advance();

        getNumberSequence();
        tokenRef.token = new Token(TokenTypes.REAL_CONST, number);
      });
    });

    return token.token;
  }

  /**
   * Moves the position of the reading pointer forward and assigns the current character in "currentCharacter".
   * in "currentCharacter".
   */
  private advance() {
    this.carretPosition += 1;

    if (this.carretPosition > this.code.length - 1) {
      this.currentCharacter = undefined;
    } else {
      this.currentCharacter = this.code[this.carretPosition];
    }
  }

  /**
   * A sort of buffer that looks for the next character without advancing the global pointer.
   * The purpose is to be able to assemble a set of characters to form a specific token.
   * Ex: When it encounters ':' it will check if the next character is '=' to form the token ':='.
   */
  private peek(): string | undefined {
    const peekPosition = this.carretPosition + 1;
    if (peekPosition > this.code.length - 1) {
      return undefined;
    }
    return this.code[peekPosition];
  }

  /**
   * Handles validation of whether the character, or sequence of characters, is a
   * reserved word or a new variable.

   * If it is a reserved word it returns the corresponding token.
   * If it is a new variable it returns the token of type ID.
   */
  private id(): Token {
    let result = '';

    while (
      !!this.currentCharacter &&
      alphanumeric.includes(this.currentCharacter)
    ) {
      result += this.currentCharacter;
      this.advance();
    }
    const type = Keywords[result] || TokenTypes.ID;
    return new Token(type, result);
  }

  /**
   * It will ignore comments for the code.
   * In Pascal comments are all code between braces.
   * Ex: {comment in pascal}
   */
  private skipComment() {
    while (this.currentCharacter !== '}') {
      this.advance();
    }
    this.advance(); // The closing curly brace
  }

  /**
   * Will ignore whitespace by moving the reading pointer forward.
   */
  private skipWhitespace() {
    while (!this.currentCharacter || this.currentCharacter === ' ') {
      this.advance();
    }
  }

  /**
   * Will ignore whitespace by moving the reading pointer forward.
   */
  private skipNewLine() {
    while (this.currentCharacter === '\n') {
      this.advance();
    }
  }

  /**
   * Returns an error if an invalid character is detected.
   */
  private raiseError(message?: string) {
    throw message || 'Invalid character detected';
  }

  /**
   * Returns a code snippet to be displayed in the error.
   */
  private getSnippet(
    initialPos = this.carretPosition - 10,
    finalPos = this.carretPosition + 10,
  ) {
    return this.code.substring(initialPos, finalPos);
  }
}
